File: Blog.Core/Infrastructure/FileProvider.cs
namespace Blog.Core.Infrastructure
{
    using Microsoft.AspNetCore.Hosting;
    using Microsoft.Extensions.FileProviders;
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Text;
    using System.Threading;
    public class FileProvider : PhysicalFileProvider, IShareFileProvider
    {
        public FileProvider(IHostingEnvironment hostingEnvironment)
            : base(File.Exists(hostingEnvironment.WebRootPath) ? Path.GetDirectoryName(hostingEnvironment.WebRootPath) : hostingEnvironment?.WebRootPath)
        {
            var path = hostingEnvironment.ContentRootPath ?? string.Empty;
            if (File.Exists(path))
                path = Path.GetDirectoryName(path);

            BaseDirectory = path;
        }

        #region Utilities

        private static void DeleteDirectoryRecursive(string path)
        {
            Directory.Delete(path, true);
            const int maxIterationToWait = 10;
            var curIteration = 0;

            //according to the documentation(https://msdn.microsoft.com/ru-ru/library/windows/desktop/aa365488.aspx) 
            //System.IO.Directory.Delete method ultimately (after removing the files) calls native 
            //RemoveDirectory function which marks the directory as "deleted". That's why we wait until 
            //the directory is actually deleted. For more details see https://stackoverflow.com/a/4245121
            while (Directory.Exists(path))
            {
                curIteration += 1;
                if (curIteration > maxIterationToWait)
                    return;
                Thread.Sleep(100);
            }
        }

        #endregion

        public virtual string Combine(params string[] paths)
        {
            return Path.Combine(paths);
        }

        public virtual void CreateDirectory(string path)
        {
            if (!DirectoryExists(path))
                Directory.CreateDirectory(path);
        }

        public virtual void CreateFile(string path)
        {
            if (FileExists(path))
                return;

            //we use 'using' to close the file after it's created
            using (File.Create(path))
            {
            }
        }

        public void DeleteDirectory(string path)
        {
            if (string.IsNullOrEmpty(path))
                throw new ArgumentNullException(path);

            //find more info about directory deletion
            //and why we use this approach at https://stackoverflow.com/questions/329355/cannot-delete-directory-with-directory-deletepath-true

            foreach (var directory in Directory.GetDirectories(path))
            {
                DeleteDirectory(directory);
            }

            try
            {
                DeleteDirectoryRecursive(path);
            }
            catch (IOException)
            {
                DeleteDirectoryRecursive(path);
            }
            catch (UnauthorizedAccessException)
            {
                DeleteDirectoryRecursive(path);
            }
        }

        public virtual void DeleteFile(string filePath)
        {
            if (!FileExists(filePath))
                return;

            File.Delete(filePath);
        }

        public virtual bool DirectoryExists(string path)
        {
            return Directory.Exists(path);
        }

        public virtual void DirectoryMove(string sourceDirName, string destDirName)
        {
            Directory.Move(sourceDirName, destDirName);
        }

        public virtual IEnumerable<string> EnumerateFiles(string directoryPath, string searchPattern,
            bool topDirectoryOnly = true)
        {
            return Directory.EnumerateFiles(directoryPath, searchPattern,
                topDirectoryOnly ? SearchOption.TopDirectoryOnly : SearchOption.AllDirectories);
        }

        public virtual void FileCopy(string sourceFileName, string destFileName, bool overwrite = false)
        {
            File.Copy(sourceFileName, destFileName, overwrite);
        }

        public virtual bool FileExists(string filePath)
        {
            return File.Exists(filePath);
        }

        public virtual long FileLength(string path)
        {
            if (!FileExists(path))
                return -1;

            return new FileInfo(path).Length;
        }

        public virtual void FileMove(string sourceFileName, string destFileName)
        {
            File.Move(sourceFileName, destFileName);
        }

        public virtual string GetAbsolutePath(params string[] paths)
        {
            var allPaths = paths.ToList();
            allPaths.Insert(0, Root);

            return Path.Combine(allPaths.ToArray());
        }

        public virtual DateTime GetCreationTime(string path)
        {
            return File.GetCreationTime(path);
        }

        public virtual string[] GetDirectories(string path, string searchPattern = "", bool topDirectoryOnly = true)
        {
            if (string.IsNullOrEmpty(searchPattern))
                searchPattern = "*";

            return Directory.GetDirectories(path, searchPattern,
                topDirectoryOnly ? SearchOption.TopDirectoryOnly : SearchOption.AllDirectories);
        }

        public virtual string GetDirectoryName(string path)
        {
            return Path.GetDirectoryName(path);
        }

        public virtual string GetDirectoryNameOnly(string path)
        {
            return new DirectoryInfo(path).Name;
        }

        public virtual string GetFileExtension(string filePath)
        {
            return Path.GetExtension(filePath);
        }

        public virtual string GetFileName(string path)
        {
            return Path.GetFileName(path);
        }

        public virtual string GetFileNameWithoutExtension(string filePath)
        {
            return Path.GetFileNameWithoutExtension(filePath);
        }

        public virtual string[] GetFiles(string directoryPath, string searchPattern = "", bool topDirectoryOnly = true)
        {
            if (string.IsNullOrEmpty(searchPattern))
                searchPattern = "*.*";

            return Directory.GetFiles(directoryPath, searchPattern,
                topDirectoryOnly ? SearchOption.TopDirectoryOnly : SearchOption.AllDirectories);
        }

        public virtual DateTime GetLastAccessTime(string path)
        {
            return File.GetLastAccessTime(path);
        }

        public virtual DateTime GetLastWriteTime(string path)
        {
            return File.GetLastWriteTime(path);
        }

        public virtual DateTime GetLastWriteTimeUtc(string path)
        {
            return File.GetLastWriteTimeUtc(path);
        }

        public virtual string GetParentDirectory(string directoryPath)
        {
            return Directory.GetParent(directoryPath).FullName;
        }

        public virtual bool IsDirectory(string path)
        {
            return DirectoryExists(path);
        }

        public virtual string MapPath(string path)
        {
            path = path.Replace("~/", string.Empty).TrimStart('/').Replace('/', '\\');
            return Path.Combine(BaseDirectory ?? string.Empty, path);
        }

        public virtual byte[] ReadAllBytes(string filePath)
        {
            return File.Exists(filePath) ? File.ReadAllBytes(filePath) : new byte[0];
        }

        public virtual string ReadAllText(string path, Encoding encoding)
        {
            return File.ReadAllText(path, encoding);
        }

        public virtual void SetLastWriteTimeUtc(string path, DateTime lastWriteTimeUtc)
        {
            File.SetLastWriteTimeUtc(path, lastWriteTimeUtc);
        }

        public virtual void WriteAllBytes(string filePath, byte[] bytes)
        {
            File.WriteAllBytes(filePath, bytes);
        }

        public virtual void WriteAllText(string path, string contents, Encoding encoding)
        {
            File.WriteAllText(path, contents, encoding);
        }

        protected string BaseDirectory { get; }
    }
}

File: Shared/Blog.FSharp.Core/Infrastructure/FileProvider.fs
namespace Blog.FSharp.Core.Infrastructure

open System
open System.IO
open System.Text
open System.Threading
open Microsoft.AspNetCore.Hosting
open Microsoft.Extensions.FileProviders

type FileProvider(hostingEnvironment: IHostingEnvironment) =
    inherit PhysicalFileProvider(
        if String.IsNullOrEmpty(hostingEnvironment.WebRootPath) then hostingEnvironment.ContentRootPath
        else hostingEnvironment.WebRootPath
    )

    let baseDirectory =
        if String.IsNullOrEmpty(hostingEnvironment.ContentRootPath) then String.Empty
        else hostingEnvironment.ContentRootPath

    member _.BaseDirectory = baseDirectory

    interface IShareFileProvider with
        member _.Combine(paths: string[]) =
            Path.Combine(paths)

        member _.CreateDirectory(path: string) =
            if not (Directory.Exists(path)) then
                Directory.CreateDirectory(path) |> ignore

        member _.CreateFile(path: string) =
            if not (File.Exists(path)) then
                use _ = File.Create(path)
                ()

        member this.DeleteDirectory(path: string) =
            if String.IsNullOrEmpty(path) then
                raise (ArgumentNullException("path"))

            let provider = this :> IShareFileProvider

            for dir in Directory.GetDirectories(path) do
                provider.DeleteDirectory(dir)

            let rec deleteDirRecursive (p: string) maxRetry =
                let mutable retry = 0
                while Directory.Exists(p) && retry < maxRetry do
                    try
                        Directory.Delete(p, true)
                    with
                    | :? IOException
                    | :? UnauthorizedAccessException ->
                        Thread.Sleep(100)
                        retry <- retry + 1

            deleteDirRecursive path 10

        member _.DeleteFile(path: string) =
            if File.Exists(path) then File.Delete(path)

        member _.DirectoryExists(path: string) = Directory.Exists(path)

        member _.FileExists(path: string) = File.Exists(path)

        member _.FileLength(path: string) =
            if File.Exists(path) then FileInfo(path).Length else -1L

        member _.MapPath(path: string) =
            path.Replace("~/", "").TrimStart('/').Replace('/', Path.DirectorySeparatorChar)
            |> fun p -> Path.Combine(baseDirectory, p)

        member _.ReadAllText(path: string, encoding: Encoding) =
            File.ReadAllText(path, encoding)

        member _.WriteAllText(path: string, contents: string, encoding: Encoding) =
            File.WriteAllText(path, contents, encoding)